### ISSUE TITLE
Every few days vagrant can't connect to vmware rest API, vmrest process becomes a zombie

### ISSUE THREAD

**Issue by stolowski** [hashicorp/vagrant#11955]
### Vagrant version

2.2.10

VMware Fusion 12

### Host operating system

Mac OS 10.15.7 (19H2)

### Guest operating system

Linux, Windows 10.

### Vagrantfile

N/A

### Debug output

N/A

### Expected behavior

vagrant commands (up, ssh ...) work all the time.

### Actual behavior

Every few days vagrant commands stop working and error out accessing vmware REST API. I can confirm vmrest API is not available by connecting with a web browser at the given http port (when it works, then I can browse these APIs with a web browser).

When this happens, vmrest process appears to be a zombie:

$ ps aux|grep vmrest                                                  
root             50024   0,0  0,0        0      0   ??  Z     6:11     0:00.00 (vmrest)

Normally (when it works, it looks like this:
$ ps aux|grep vmrest
root               344   0,0  0,1  5316588  34984   ??  S     6:23     0:00.33 /Applications/VMware Fusion.app/Contents/Library/vmrest

### Steps to reproduce

No clear steps. Happens randomly every few days, I keep my Mac running for several days and only close the lid overnight. I don't think it is strictly related to waking it from sleep in the morning, it also happened close to end of my work day.

### References

I've created a VMware support ticket as well, but it is at early stages yet.

I suspect the problem **might be** with VMware/vmrest; on the other hand it seems that vmrest is started on demand (triggered / spawned by vagrant commands), and if vmrest becomes a zombie, it may suggest it's parent process quit abruptly. I'd appreciate any thoughts or ideas on debugging this, or any suggestions about how to recover (e.g. if restarting any daemons would help), as at the moment, in order to recover and have vagrant operational again I have to restart my system whenever this happens. I think this issue started happening after upgrading to Fusion 12 (and upgrading vagrant for it).

**Comment by chrisroberts** [comment]
@stolowski I'll be looking to reproduce this behavior today and hopefully identify and resolve the issue fairly quickly.

**Comment by stolowski** [comment]
> @stolowski I'll be looking to reproduce this behavior today and hopefully identify and resolve the issue fairly quickly.

Thank you! To be honest, I'm a little skeptical when it comes to reproducing it in a short time; it just happened here again after a few days without problems, and after a few hours after of working today without experiencing the problem. I'm attaching vagrant debug in case it shows anything that may give a clue...

If you have any suggestions where to look for more logs/debug, please let me know.

[vagrant.log](https://github.com/hashicorp/vagrant/files/5395808/vagrant.log)

**Comment by stolowski** [comment]
This just happened again, so 5 days from previous occurence. After restarting the system I took a look at the processes with pstree and under normal conditions I see this:
```
 |-+= 00318 root /opt/vagrant-vmware-desktop/bin/vagrant-vmware-utility api -port=9922 -config-file=/opt/vagrant-vmware-desktop/config/service.hcl
 | \--- 02980 root /Applications/VMware Fusion.app/Contents/Library/vmrest
```

which means vagrant-vmware-utility is in control of vmrest. Is there a service related to vagrant-vmware-utility that I could restart as a workaround when this happens?

**Comment by stolowski** [comment]
@chrisroberts I had a hunch and checked Mac's Console app and found two crash reports for vagrant-vmware-utility
 that match the times where vmrest wouldn't work anymore; attached below.

[crash1.txt](https://github.com/hashicorp/vagrant/files/5421347/crash1.txt)
[crash2.txt](https://github.com/hashicorp/vagrant/files/5421348/crash2.txt)

**Comment by stolowski** [comment]
This just happened again. I was hoping `sudo launchctl start  com.vagrant.vagrant-vmware-utility` would work around the problem, but it didn't help (the service would start, but vmrest does not respond). @chrisroberts Is there any workaround you can think of that I could try next time? Any logs or other information I should get when it happens again? This is a very frustrating problem...

**Comment by stolowski** [comment]
I know recent update to vagrant/vagrant-vmware-utility was aimed at fixing the issue of zombie vmrest process and I think it indeed fixed it, but I just experienced the problem again - this time there there was no vmrest zombie process, vagrant-vmware-utility was running, but it didn't not (re)start vmrest, so I was getting the usual "Get http://localhost:49743/api/vmnet: GET http://localhost:49743/api/vmnet giving up after 5 attempts" error from vagrant commands.

This time however I was able to recover with:
```
pawel@macbook-pro ~/Vagrant/ubuntu-20.04-snapd (master*) $ sudo launchctl stop com.vagrant.vagrant-vmware-utility
pawel@macbook-pro ~/Vagrant/ubuntu-20.04-snapd (master*) $ sudo launchctl start com.vagrant.vagrant-vmware-utility
```
which also started vmrest.

So I think this is going in the right direction but it would be great if com.vagrant.vagrant-vmware-utility service automatically restarted vmrest if needed.

**Comment by chrisroberts** [comment]
@stolowski had you put the machine to sleep between the time it was functioning correctly, and when it began to error with the vmrest process not being started?

Also, if you could toss in the logs from `/Library/Application Support/vagrant-vmware-utility/service.log` around when you experience this, that would be great. I'll work on trying to repro this again with the latest utility.

Thanks!

**Comment by stolowski** [comment]
@chrisroberts No, I didn't put the machine to sleep around the time the issue occurred. I woke it in the morning around 9:00, started a vm with `vagrant up` followed by `vagrant ssh`, then worked on it over ssh for a couple of hours (without issuing any vagrant commands for around 4 hours which seems to be confirmed by the log). The problem occurred around 13:09 when I tried to issue a vagrant command.

Attaching the log for that day, I think you should be looking for 2020/11/13 13:09.
[service.log.gz](https://github.com/hashicorp/vagrant/files/5540555/service.log.gz)

(there are a couple of vagrant-vmware-utility restarts after that time which are my attempts to recover and some playing with automatic the recovery).

**Comment by uberjay** [comment]
Ah, I've been running into this for quite some time now, so it's helpful to see what the cause is. I'd gotten so far as figuring out that the vmrest was in zombie state (a couple months ago), but assumed it was a VMware bug and didn't have any more time to spend tracking it down.

This time, when it failed, I was able to restore vagrant vmware rest api service by restarting `com.vagrant.vagrant-vmware-utility` with `launchctl` -- thanks for that!

If there's any additional debug info you might need, I could poke around as well.

**Comment by stolowski** [comment]
FWTW I haven't experienced this problem with vagrant 2.2.14 for good 3 weeks, but I've also upgraded to Big Sur on the day it was officially released (on a side note, I don't recommend switching to Big Sur without doing some research/reading first, i.e. there are serious issues with nested virtualization with Fusion 12 on it). All my previous problems related to that issue were from Catalina.

**Comment by kernelsmith** [comment]
This has been happening to me and those on my team for years, at least as far back as this issue report. This happens so often that we have a shell script that runs:

```Shell
sudo launchctl unload -w /Library/LaunchDaemons/com.vagrant.vagrant-vmware-utility.plist
sudo launchctl load -w /Library/LaunchDaemons/com.vagrant.vagrant-vmware-utility.plist
```

**Comment by uberjay** [comment]
> This has been happening to me and those on my team for years, at least as far back as this issue report. This happens so often that we have a shell script that runs:

Yep, me too. It's just the normal part of the routine for me. Eventually, routine will transition into tradition. It'll feel good and right and true, but we'll have long since forgotten why we're doing it!! 😁

**Comment by defanator** [comment]
Hi. Seeing this with vagrant 2.4.9 on macOS 15.7.5 and vmware fusion 25H2u1.

vagrant-vmware-utility version is 1.0.24:
```
% /opt/vagrant-vmware-desktop/bin/vagrant-vmware-utility --version
1.0.24
```

When looking into "/Library/Application Support/vagrant-vmware-utility/service.log" during the stage when things are broken, I spotted these events:
```
2026-05-06T13:36:51.118-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request start: method=GET path=/vmware/info request-id=0x140001fa380
2026-05-06T13:36:51.118-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request complete: code=200 request-id=0x140001fa380
2026-05-06T13:36:51.122-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request start: method=GET path=/vmware/paths request-id=0x140001fa460
2026-05-06T13:36:51.122-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request complete: code=200 request-id=0x140001fa460
2026-05-06T13:36:51.123-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output="VMware Fusion REST API"
2026-05-06T13:36:51.123-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output="Copyright (C) 2015-2026 Broadcom."
2026-05-06T13:36:51.123-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output="All Rights Reserved"
2026-05-06T13:36:51.123-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output=""
2026-05-06T13:36:51.123-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output="vmrest 1.3.1 25219963"
2026-05-06T13:36:51.124-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request start: method=GET path=/version request-id=0x140001fa540
2026-05-06T13:36:51.124-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request complete: code=200 request-id=0x140001fa540
2026-05-06T13:36:51.124-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output="To listen on TCP port, Please use -C to update credential"
2026-05-06T13:36:51.126-0600 [INFO]  vagrant-vmware-utility.api.api.handler: request start: method=GET path=/vmnet request-id=0x140003108c0
2026-05-06T13:36:51.126-0600 [INFO]  vagrant-vmware-utility.api.vmrest: starting remote request to vmware service
2026/05/06 13:36:51 [DEBUG] GET http://localhost:49165/api/vmnet
2026/05/06 13:36:51 [ERR] GET http://localhost:49165/api/vmnet request failed: Get "http://localhost:49165/api/vmnet": dial tcp [::1]:49165: connect: connection refused
2026/05/06 13:36:51 [DEBUG] GET http://localhost:49165/api/vmnet: retrying in 1s (4 left)
2026-05-06T13:36:51.826-0600 [INFO]  vagrant-vmware-utility.api.vmrest.process: vmrest stdout: output="Not listening to either Unix Socket or TCP port, exiting"
2026-05-06T13:36:51.828-0600 [WARN]  vagrant-vmware-utility.api.vmrest.process: stdout pipe error: error=EOF
2026-05-06T13:36:51.828-0600 [WARN]  vagrant-vmware-utility.api.vmrest.process: stderr pipe error: error=EOF
```

It is happening once in a few days without any noticeable reasons.

What those credentials from `vmrest -C` are about? Somehow these got stale or disappear over time?

**Comment by defanator** [comment]
I know it's been a while, but https://github.com/hashicorp/vagrant-vmware-desktop/pull/175 seems to fix the issue in my affected environments. If any of you folks are still around, I'd appreciate some testing and maybe some voices as it seems like nobody is looking after https://github.com/hashicorp/vagrant-vmware-desktop nowadays.

@stolowski @chrisroberts @uberjay @kernelsmith
